desktop/cmp/persistenceManager/PersistenceManager.ts

```diff
@@ -0,0 +1,214 @@
+                saveDialogModel ? saveDialog({key: saveDialogModel.xhId}) : null
+            );
+        }
+    });
+
+//------------------------
+// Implementation
+//------------------------
+
+const saveButton = hoistCmp.factory<PersistenceManagerModel>({
+    render({model}) {
+        return button({
+            icon: Icon.save(),
+            tooltip: `Save changes to this ${model.noun}`,
+            intent: 'primary',
+            omit: !model.enableTopLevelSaveButton || !model.canSave,
+            onClick: () => model.saveAsync(null).linkTo(model.loadModel)
+        });
+    }
+});
+
+const objMenu = hoistCmp.factory<PersistenceManagerModelProps>({
+    render({model, minimal}) {
+        const {pluralNoun, objects, loadModel} = model,
+            grouped = groupBy(objects, it => it.group),
+            sortedGroupKeys = keys(grouped).sort(),
+            items = [];
+
+        sortedGroupKeys.forEach(group => {
+            items.push(menuDivider({title: group}));
+            items.push(...hierarchicalMenus(sortBy(grouped[group], 'name')));
+        });
+
+        return menu({
+            items: [
+                ...items,
+                fragment({
+                    omit: minimal,
+                    items: [
+                        menuDivider(),
+                        menuItem({
+                            icon: Icon.plus(),
+                            text: 'New...',
+                            onClick: () => model.createNewAsync().linkTo(loadModel)
+                        }),
+                        menuItem({
+                            icon: Icon.save(),
+                            text: 'Save',
+                            disabled: !model.canSave,
+                            onClick: () => model.saveAsync(null).linkTo(loadModel)
+                        }),
+                        menuItem({
+                            icon: Icon.copy(),
+                            text: 'Save as...',
+                            onClick: () => model.saveAsAsync().linkTo(loadModel)
+                        }),
+                        menuItem({
+                            icon: Icon.reset(),
+                            text: 'Reset',
+                            disabled: !model.isDirty,
+                            onClick: () => model.resetAsync().linkTo(loadModel)
+                        }),
+                        menuDivider(),
+                        menuItem({
+                            icon: Icon.gear(),
+                            text: `Manage ${pluralNoun}...`,
+                            onClick: () => model.openManageDialog()
+                        })
+                    ]
+                })
+            ]
+        });
+    }
+});
+
+/**
+ * @param records
+ * @param depth  used during recursion, depth in the path string/hierarchy
+ * @returns an array of menuItem()s
+ */
+function hierarchicalMenus(records: PlainObject[], depth: number = 0): ReactNode[] {
```

Comment:
Interesting to think about whether a function like this should live at the component or model level.  Seems like this could be a method in the model that returns an object representing a menu structure, and that's what could be passed to a component capable of rendering it.